add TypeBox to add and remove boxed/runtime-typed values

Jakobeha opened the pull request:
adds `TypeBox` which is simply a wrapper for `Box<dyn Any + Send + Sync>` which prevents the need for `as` casts (could also do `type TypeBox = Box<dyn Any + Send + Sync>`), and adds the ability to add and remove `TypeBox` directly so that you can don't need to know the object's type at compile time to add or remove it from the map.

Jakobeha committed:
add TypeBox to add and remove boxed values

Jakobeha committed:
fixed bugs

Jakobeha committed:
remove warning

DoumanAsh commented on src/lib.rs:
```diff
@@ -68,6 +68,11 @@ pub struct TypeMap {
+///Boxed dynamically-typed value
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct TypeBox(pub Box<dyn core::any::Any + Send + Sync>);
```
use `Value` type alias
Make it public and rename to `PtrValue`

DoumanAsh commented on src/lib.rs:
```diff
@@ -184,6 +189,23 @@ impl TypeMap {
 
+    ///Insert element inside the map via its dynamic type-id,
+    ///returning heap-allocated old one with the same type-id if any
+    pub fn insert_box(&mut self, value: TypeBox) -> Option<TypeBox> {
```
`insert_box`/`remove_box` -> `_raw`

DoumanAsh reviewed (changes requested):
This seems to be for overly specialized use case?
I'm not sure why you generally would need it, but I do not mind to add it
In that case please also add getter for type erased variant

Jakobeha committed:
rename Type to Value and Value to ValueBox, add getters for ref and mut, no TypeBox wrapper

Jakobeha commented:
I applied some of your changes, except used `ValueBox` instead of `PtrValue`. I also renamed `Type` to `Value` because I thought it made more sense esp. with `ValueBox`. And added the `ValueRef` and `ValueMut` types.

It's not a big deal if you don't agree to these changes, as I can just use the forked version in my projects

Jakobeha committed:
formatting

DoumanAsh commented on src/lib.rs:
```diff
@@ -57,16 +56,23 @@ macro_rules! unreach {
-pub trait Type: 'static + Send + Sync {}
-impl<T: 'static + Send + Sync> Type for T {}
+///Valid type for [TypeMap]
+pub trait Value: 'static + Send + Sync {}
```
Semantically I would prefer for it to remain `Type`
Please note that this is public interface, while I do not mind to bump semver, there is no real need to rename Type to Value

DoumanAsh commented:
Published 0.2.1
